[Governance] Proof of Fee

0o-de-lally opened the pull request:
Implements the first version of Proof of Fee as per the paper  https://0l.network/2022/10/15/proof-of-fee-part-1/.
Work started after the community poll: https://0l.network/2022/11/08/october-2022-governance-recap/

- [x] Initialize bids, and expiry.
- [x] Functions for auction: Fill the seats and determine the clearing price
- [x] Near complete refactor of epoch boundary
  - [x] Elect validators based on auction
  - [x] Separate Jail Mechanics
- [x] Remove fee calcluator, for hardcoded Baseline Fee. 
- [x] Functions for charging arbitrary Fees  to the epoch's transaction fee account.
- [x] Charge Fee on Admission, goes to transaction fee account
- [x] Pay validators Baseline fee from chain fee account
- [x] Burn remainder in transaction fee account
- [x] limits on bids: % above baseline, unlocked in account
- [x] Change the weight of consensus
- [x] Baseline fee ratcheting based on bids
- [x] Future proof for case where we want to limit bid retractions
- [ ] Stretch Goal: Priority seating based on Jail reputation

Remove Stuff
- [x] Remove Cases, towers don't apply for validator compliance
- [x] Keep TowerState, but remove validator privileges from Proof of Height
- [x] NodeWeight

Transactional Tests
- [x] Make unrelated tests pass
- [x] Bid setting and initialization
- [x] Epoch Boundary makes progress
- [x] Validator election tests
- [x] Correct Fee payments based on compliance
- [x] Vouching tests
- [x] Thermostat adjustment tests
- [x] Median calc tests
- [x] Burn excess fees test
- [x] Testnet Validator creation sanity tests

0o-de-lally committed:
For 0L genesis tools for hard forks, and migrations: create a MoveVm session from fake data to be able to apply migrations from Move system contracts.

0o-de-lally committed:
Scaffolding tests for v6 genesis fork. first cleanup deprecated modules.

0o-de-lally committed:
add fixtures for testing recovery genesis

0o-de-lally committed:
cleanup fixtures
implement functional tests for:
- exporting db backup to json
- creating genesis blob from json
- creating genesis blob in one shot from db backup

TODO: launch test node from from genesis.blob
TODO: e2e tests from a fixture file

0o-de-lally committed:
integration test using blob from fixture to start a diem-node in test mode

0o-de-lally committed:
patch github actions tests

0o-de-lally committed:
full e2e test from db backup to starting a node

0o-de-lally committed:
PoF requires all nodes to have equal weight in consensus per the paper.

0o-de-lally committed:
initial scaffold of proof-of-fee auction. Borrows code from NodeWeight. Includes tranaction api.

0o-de-lally committed:
implement seat filling algo, which considers the 1/3 unproven node limit on epoch changes

0o-de-lally committed:
patch

0o-de-lally committed:
patch

0o-de-lally committed:
builds

0o-de-lally committed:
Add diem_logger as per @corythian plus comments.

0o-de-lally committed:
query nodes to check if onchain state is the same as recovery.

0o-de-lally committed:
wip testing fakeexector for account migrations

0o-de-lally committed:
- include db reader util to inspect a genesis.blob.
- refactor recovery tests

0o-de-lally committed:
test to check validator count. cleanup test utlis

0o-de-lally committed:
cleanup tests

0o-de-lally committed:
stub a "check" feature, to check a json file agains the output genesis.blob

0o-de-lally committed:
finish balance comparison with error list.

0o-de-lally committed:
clean

0o-de-lally committed:
test to create single validator and check output.

0o-de-lally committed:
skip comparing system zero address.

0o-de-lally committed:
cargo fix

0o-de-lally committed:
cargo fmt

0o-de-lally committed:
create function to start a vm executor, and uses a function pointer to transform legacy data to Move arguments.

0o-de-lally committed:
experimenting with alternative way of starting a vm genesis session

0o-de-lally committed:
runs

0o-de-lally committed:
cleanup

0o-de-lally committed:
need to explicitly set the validators at genesis

0o-de-lally committed:
test to confirm genesis validator set, is the same as defined set.

0o-de-lally committed:
tests for validator set and balance passing

0o-de-lally committed:
change fixtures, make node start from command line

0o-de-lally committed:
can start a test node from genesis blob

0o-de-lally committed:
can genesis tools test can start a node from a genesis and query for an accounts resources

0o-de-lally committed:
make ol version of swarm config generator in ol/smoke_tests

0o-de-lally committed:
wip

0o-de-lally committed:
implement vm function to pay fee from user account

0o-de-lally committed:
all validators are charged fee at epoch boundary

0o-de-lally committed:
cleanup  validator set proposal functions

0o-de-lally committed:
refactor jail function at epoch boundary

0o-de-lally committed:
make epoch boundry _meta_epoch.move pass

0o-de-lally committed:
filter for vouches on proof of fee

0o-de-lally committed:
initialize val account with a PoF struct

0o-de-lally committed:
epoch reconfiguration meta tests passing. PoF bid initialization unit tests passing. Validator meta sanity tests passing

0o-de-lally committed:
patch issue with failover

0o-de-lally committed:
add bid restrictions, and checking upon seating

0o-de-lally committed:
burn excess fees than needed for validators

0o-de-lally committed:
reward thermostat initial implementation

0o-de-lally committed:
move legacyRecovery to ol/types. So now we can run the user migrations in the vm_genesis::lib. First draft builds

0o-de-lally committed:
create GenesisMigration user migration test. Builds.

0o-de-lally committed:
ol/genesis-tools builds with refactor

0o-de-lally committed:
refactored with clean build

0o-de-lally committed:
minor refactors

0o-de-lally committed:
can migrate the balance of end users with new design

0o-de-lally committed:
migration successfully migrates balances of users, validators, and operators.

0o-de-lally committed:
genesis init fee baseline

0o-de-lally committed:
patch bug in thermostat

0o-de-lally committed:
more tests for fill_seats, and thermostat. patches

0o-de-lally committed:
test

0o-de-lally committed:
consolidate validator audits, check for the case that there are fewer bidders than the stated val set.

0o-de-lally committed:
add test fixtures, modify some APIs "tell don't ask" pattern. Add several tests.

0o-de-lally committed:
add unit tests for thermostat. Patch a couple bugs

0o-de-lally committed:
tests for thermostat increase/decrease

0o-de-lally committed:
deprecate Audit.move, and create tests for qualifications in Proofoffee.move

0o-de-lally committed:
Jailing mechanism needs update after deprecating Towers. Proposing here that a Voucher is the only one to unjail, since there is no other practical hurdle.

0o-de-lally committed:
audit tests for vouch, expired bids, no_funds

0o-de-lally committed:
covered all proof of fee functions, tests for all common scenarios of fill_seats, patch logic issues, all proof of fee tests passing!

0o-de-lally committed:
patch seating failover

0o-de-lally committed:
Clean up Cases.move. Patch Vouch tests fixtures

0o-de-lally committed:
adjust fixtures for reconfiguration tests, deprecate some tests, patch jailing

0o-de-lally committed:
patch all reconfig test fixtures. remove deprecated burn tests.

0o-de-lally committed:
continue patching tests and removing deprecated scenarios

0o-de-lally committed:
All transactional tests passing!

0o-de-lally committed:
cleanup unused files. clippy,

0o-de-lally committed:
tests passing

0xzoz committed:
Update ci.yaml (#1239)

hemulin commented on diem-move/diem-framework/DPN/sources/DiemAccount.move:
```diff
@@ -546,6 +548,20 @@ module DiemFramework::DiemAccount {
+        // value: u64,
+    ) acquires AccountOperationsCapability  {
+        CoreAddresses::assert_diem_root(vm);
+        // TODO: fix the timestamp issue
```
@0o-de-lally can you comment on the timestamp issue? There is no common timestamp in the form of height since genesis?

0o-de-lally replied:
There's a bug in this version of Diem. @Corynthian had pointed it out in another issue. Would like to pull that change into V6. Doesn't have to be on this branch.

hemulin reviewed (commented):
I intend to pass in multiple rounds.

General question, what is the motive behind leaving much commented-out code?

0o-de-lally commented:
> General question, what is the motive behind leaving much commented-out code?
People may be reading/reviewing outside of the github tools, and I want to be explicit about the big changes. I'll remove when ready for merge.

0o-de-lally committed:
txs transaction apis for setting and retracting bids.

0o-de-lally committed:
test for set and retract

0o-de-lally committed:
[Genesis] Migration tools for genesis from legacy accounts (#1237)

* For 0L genesis tools for hard forks, and migrations: create a MoveVm session from fake data to be able to apply migrations from Move system contracts.

* wip testing fakeexector for account migrations

* Scaffolding tests for v6 genesis fork. first cleanup deprecated modules.

* add fixtures for testing recovery genesis

* cleanup fixtures
implement functional tests for:
- exporting db backup to json
- creating genesis blob from json
- creating genesis blob in one shot from db backup

TODO: launch test node from from genesis.blob
TODO: e2e tests from a fixture file

* integration test using blob from fixture to start a diem-node in test mode

* patch github actions tests

* full e2e test from db backup to starting a node

* Add diem_logger as per @corythian plus comments.

* query nodes to check if onchain state is the same as recovery.

* - include db reader util to inspect a genesis.blob.
- refactor recovery tests

* test to check validator count. cleanup test utlis

* cleanup tests

* stub a "check" feature, to check a json file agains the output genesis.blob

* finish balance comparison with error list.

* clean

* test to create single validator and check output.

* skip comparing system zero address.

* cargo fix

* cargo fmt

* create function to start a vm executor, and uses a function pointer to transform legacy data to Move arguments.

* experimenting with alternative way of starting a vm genesis session

* runs

* cleanup

* need to explicitly set the validators at genesis

* test to confirm genesis validator set, is the same as defined set.

* tests for validator set and balance passing

* change fixtures, make node start from command line

* can start a test node from genesis blob

* can genesis tools test can start a node from a genesis and query for an accounts resources

* make ol version of swarm config generator in ol/smoke_tests

* wip

* move legacyRecovery to ol/types. So now we can run the user migrations in the vm_genesis::lib. First draft builds

* create GenesisMigration user migration test. Builds.

* ol/genesis-tools builds with refactor

* refactored with clean build

* minor refactors

* can migrate the balance of end users with new design

* migration successfully migrates balances of users, validators, and operators.

* cleanup unused files. clippy,

* tests passing

* whitespace to trigger ci

sirouk commented:
LGTM, thanks!

0o-de-lally committed:
Merge branch 'v6' into v6-0d-proof-of-fee

0o-de-lally commented:
@hemulin  Reviewers should focus on the changed files in: `diem-move/diem-framework/DPN/sources/0L/`

Perhaps starting with:
`diem-move/diem-framework/DPN/sources/0L/ProofOfFee.move`
Then
`diem-move/diem-framework/DPN/sources/0L/EpochBoundary.move` and walk from there.

0o-de-lally committed:
Merge branch 'v6' into v6-0d-proof-of-fee

0o-de-lally committed:
cleanup prints

0o-de-lally committed:
add struct in TransactionFee to track origin of fees FeeMaker

0o-de-lally committed:
refactor Burn to consider the preference of the user FeeMaker

0o-de-lally committed:
cleanup

0o-de-lally committed:
init FeeMaker on account creation

0o-de-lally committed:
patch tests to remove `print`

0o-de-lally committed:
patch epoch boundary burn ratios

0o-de-lally committed:
all tests passing

hemulin commented:
Generally speaking, looking great :muscle: Super impressive @0o-de-lally 

The only comments I may suggest are:
1. commented-out-code cleaning. The comments are very helpful to understand the idea behind the methods, however there are plenty of commented out lines that can and should be removed before merge (all the `print(&whatever)`)

2. Error handling - Not likely to happen but this is what error handling is for.  For example, in `fill_seats_and_get_price` what will happen if `i >= Vector::length(sorted_vals_by_bid)`? The while loop will break, we fail to seat anyone and though the comment point to let `EpochBoundary` to deal with it, the EpochBoundary is again calling `fill_seats_and_get_price` which may fail again and we are in an endless cycle.

0o-de-lally commented:
@hemulin 
> Error handling - Not likely to happen but this is what error handling is for. For example, in fill_seats_and_get_price what will happen if i >= Vector::length(sorted_vals_by_bid)? The while loop will break, we fail to seat anyone and though the comment point to let EpochBoundary to deal with it, the EpochBoundary is again calling fill_seats_and_get_price which may fail again and we are in an endless cycle.

Thanks for this. I'm sure you are right that there are more places that could have error handling.
In this case specifically the job of fill seats is to tell us if it could fill seats.  An empty result is ok. Whether or not we start an epoch with that list is handled elsewhere. Take a look at EpochBoundary propose_new_set where there is a section `//////// Failover Rules ////////` which is a catch all on issues with validator selection.

So yes, let's do more on this topic. Keep the suggestions coming!

0xzoz commented:
Closing due to changes merged in #5
